### PR TITLE
Switch mozci to using pushlog_client

### PR DESCRIPTION
**Pull request by MikeLing**

### CODE REVIEW

**Comment by MikeLing** [comment]
Hmm, I meet some problem for this PR, I got error like:
============================= test session starts ==============================
platform darwin -- Python 2.7.10, pytest-2.8.2, py-1.4.30, pluggy-0.3.1
rootdir: /Users/mikeling/mozilla_ci_tools, inifile: 
collected 8 items 

test/test_mozci.py ...F....

=================================== FAILURES ===================================
____________________ TestQueries.test_query_revisions_range ____________________

self = <test_mozci.TestQueries testMethod=test_query_revisions_range>
query_pushes_by_revision_range = <MagicMock name='query_pushes_by_revision_range' id='4356127120'>

```
@patch('mozhginfo.pushlog_client.query_pushes_by_revision_range',
       return_value=MOCK_PUSH_LIST)
def test_query_revisions_range(self, query_pushes_by_revision_range):
    self.assertEqual(mozci.mozci.query_revisions_range(self.repo_name, self.start_revision,
                                                      self.end_revision),
                     ['eb15e3f893453d6a4472f8905271aba33f8b68d5',
                      '1c5b4332e2f1b73fe03977b69371e9a08503bff3',
                      '724f0a71d62171da1357e6c1f93453359e54206b'])
```

test/test_mozci.py:86: 

---

mozci/mozci.py:326: in query_revisions_range
    from_revision, to_revision)

---

repo_url = 'https://hg.mozilla.org/integration/mozilla-inbound'
from_revision = '8850aa0f93453d6a4472f8905271aba33f8b68d5'
to_revision = '2a193b7fd62171da1357e6c1f93453359e54206b', version = 2
tipsonly = 1

   ???
E   TypeError: string indices must be integers

*\* build/bdist.macosx-10.6-intel/egg/mozhginfo/pushlog_client.py:53: TypeError **
====================== 1 failed, 7 passed in 1.69 seconds ===================

But actually, I think the pushlog_client.py shouldn't been called because I had mock it and make it return a list of push log object. I don't know what's going on with it :(
could you give me some advice about it? Thank you! @adusca

File: mozci/mozci.py [hunk 1]
```diff
@@ -316,11 +322,10 @@ def query_repo_url_from_buildername(buildername):
 
 def query_revisions_range(repo_name, from_revision, to_revision):
     """Return a list of revisions for that range."""
-    return pushlog.query_revisions_range(
-        repositories.query_repo_url(repo_name),
-        from_revision,
-        to_revision,
-    )
+    pushes = query_pushes_by_revision_range(repositories.query_repo_url(repo_name),
```
Comment (armenzg) [inline mozci/mozci.py]: Could you please list the arguments of the function explicitely? Please try to learn this approach and use for all following patches. It is very important.

query_pushes_by_revision_range(
   repo_url=repositories.query_repo_url(repo_name),
   ...=from_revision,
   ...=to_revision
)
(Using ... instead of finding the argument name)

File: mozci/sources/pushlog.py [hunk 2]
```diff
@@ -37,70 +44,48 @@ def query_revisions_range(repo_url, from_revision, to_revision, version=2, tipso
...
+                                          start_id=start_id,
+                                          end_id=end_id,
+                                          version=version)
+    return _exact_revision_from_push(pushes)
 
 
-def query_revisions_range_from_revision_before_and_after(repo_url, revision, before, after):
+# The original function named query_revisions_range_from_revision_before_and_after
+def query_revisions_by_specified_revision_range(repo_url, revision, before, after):
     """
```
Comment (MikeLing) [inline mozci/sources/pushlog.py]: I change the function name from 
**query_revisions_range_from_revision_before_and_after** ---> **query_revisions_by_specified_revision_range**,   hope it's more easier to used :)

**Comment by MikeLing** [comment]
In this PR, we will keep the pushlog.py and the api inside it, because it's easy to broken something and maybe need a lot of work for delete it and use pushlog module directly.

Instead of that, we will make the pushlog.py call pushlog module inside and keep the way we use pusglog.py in other places. :)

could you take a look at this @armenzg @adusca ? Hope it will fit your mind, thank you :D

**Comment by adusca** [comment]
Hi @MikeLing! I believe it would be better if we remove pushlog.py, we can just use functions from pushlog_client directly.

**Comment by armenzg** [comment]
As Alice points out, we want to get rid of pushlog.py completely.
This work will also help make sure that all changes were moved properly to mozhginfo.

Thanks Mikieling!

**Comment by armenzg** [comment]
I released the latest version of mozhginfo.
Let me know when there are changes to review in here.

**Comment by armenzg** [comment]
I see some commits from 16 days ago; are there more changes coming or would you like me to review it?

**Comment by MikeLing** [comment]
no, I haven't test it yet :)

**Comment by MikeLing** [comment]
I just use these commands blow to test it and it works good to me :)

<pre><code>
python mozci/scripts/trigger.py --repo-name try --revision a1e682c00ff3 -b "Ubuntu VM 12.04 x64 try debug test mochitest-4" --dry-run --debug
python mozci/scripts/buildbot_to_taskcluster.py --trigger-from-task-id E8ReTsofQcujc9iSTmaR7A --repo-name try --revision c916d035c657 --builders "['Ubuntu VM 12.04 x64 try debug test mochitest-4']" --dry-run --debug
python mozci/scripts/triggerbyfilters.py mozilla-inbound tip  --includes "web-platform-tests,debug"
</code></pre>

So I ask review for it. @armenzg r?

**Comment by armenzg** [comment]
I run out of time to look at this. I will look at it on Monday.
If @xennygrimmato or @F3real have time to review it before that it would be great!

There are also some other PRs from @F3real I couldn't get to so if you can that would be superb.
If not I will see you all on Monday.

**Comment by armenzg** [comment]
Looking at this.

File: mozci/mozci.py [hunk 3]
```diff
@@ -274,6 +279,15 @@ def _find_files(job_schedule_info):
     return files
 
 
+def exact_revision_from_push(pushes):
+    """accept a list of push object and return a list of revision"""
+    revisions = []
+    for push in pushes:
+        revisions = revisions + [changeset.node for changeset in push.changesets]
```
Comment (armenzg) [inline mozci/mozci.py]: This would actually not work.

For instance, http://hg.mozilla.org/mozilla-central/pushloghtml?changeset=5acc2a44834c

Only the tip most changeset is valid for Buildapi:
https://secure.pub.build.mozilla.org/buildapi/self-serve/mozilla-central/rev/5acc2a44834c

All the other changesets cannot be used to trigger any new builds or tests.

Comment (armenzg) [inline mozci/mozci.py]: This actually works (I tested it) because json-pushes returns the same output for every changeset on a push:
https://hg.mozilla.org/try/json-pushes?changeset=a1e682c00ff3&tipsonly=1
https://hg.mozilla.org/try/json-pushes?changeset=a29f3fd54de2&tipsonly=1

Tipsonly returns tips only :)

Even if I try to use the wrong revision it gets corrected (a1e682c00ff3 instead of a29f3fd54de2):

> "POST /buildapi/self-serve/try/builders/Ubuntu%20VM%2012.04%20x64%20try%20debug%20test%20mochitest-4/a1e682c00ff34c054a1217a8f057038cc8f8ffbb HTTP/1.1" 202 58

When you call [1] query_pushes_by_revision_range() you always loose information about other changesets but the tip one.

In other words, we can't hit a bug in here because there is a bug I introduced a while ago in the pushlog module:
https://bugzilla.mozilla.org/show_bug.cgi?id=1238623

Can you please fix the code above so we only deal with tip changesets and ignore the rest?

[1] https://hg.mozilla.org/hgcustom/version-control-tools/file/default/pylib/mozhginfo/mozhginfo/pushlog_client.py#l34

File: mozci/mozci.py [hunk 4]
```diff
@@ -274,6 +279,15 @@ def _find_files(job_schedule_info):
     return files
 
 
+def exact_revision_from_push(pushes):
```
Comment (armenzg) [inline mozci/mozci.py]: Could you please change the name to "_pushes_to_list()"?

Comment (armenzg) [inline mozci/mozci.py]: I believe this function should be in mozhginfo and all the functions of mozhginfo should return their lists as lists of strings rather than lists of pushes.

The users of the client as it is are required to add a function like this into their code. They can't use the client without having to manipulate the data first.

**Comment by armenzg** [comment]
This is very well done. Thanks Mike!
Just moving that function to mozhginfo and adjust the code in here and we should be good to go.

**Comment by MikeLing** [comment]
@armenzg r? And I found the buildbot_to_taskcluster has been broken for a while. I will create a issue for it :)

File: mozci/mozci.py [hunk 5]
```diff
@@ -554,11 +560,12 @@ def manual_backfill(revision, buildername, max_revisions, dry_run=False):
...
     # We want to use data from treeherder for manual backfilling for long term.
     set_query_source("treeherder")
-    revlist = pushlog.query_revisions_range_from_revision_before_and_after(
+    revlist = query_pushes_by_specified_revision_range(
         repo_url=repo_url,
         revision=revision,
         before=max_revisions,
-        after=-1)  # We do not want the current job in the revision to be included.
+        after=-1,
+        return_revision_list=True)  # We don't want the current job in the revision to be included.
```
Comment (armenzg) [inline mozci/mozci.py]: The comment should go on the line where the -1 is.

File: mozci/scripts/trigger.py [hunk 6]
```diff
@@ -295,11 +295,11 @@ def main():
 
     repo_url = query_repo_url(repo_name)
     if options.rev == 'tip':
-        revision = query_repo_tip(repo_url)
+        revision = query_repo_tip(repo_url).changesets[0].node
         LOG.info("The tip of %s is %s", repo_name, revision)
 
     else:
-        revision = query_full_revision_info(repo_url, options.rev)
+        revision = query_push_by_revision(repo_url, options.rev).changesets[0].node
```
Comment (armenzg) [inline mozci/scripts/trigger.py]: I think we can use return_revision_list=True
https://hg.mozilla.org/hgcustom/version-control-tools/file/default/pylib/mozhginfo/mozhginfo/pushlog_client.py#l170

File: setup.py [hunk 7]
```diff
@@ -17,7 +17,8 @@
         'progressbar>=2.3',
         'requests>=2.5.1',
         'taskcluster>=0.0.28',
-        'treeherder-client>=2.0.1'
+        'treeherder-client>=1.4',
```
Comment (armenzg) [inline setup.py]: You need to merge from master to grab the latest changes.

**Comment by armenzg** [comment]
@MikeLing this is very well done. Very minor issues.
Feel free to address the issues and push back with squashing since I don't need my previous comments.

So good to see this getting this far!

**Comment by MikeLing** [comment]
@armenzg r ?

**Comment by armenzg** [comment]
\o/
